src/codegen.cpp

```diff
@@ -6610,10 +6611,29 @@ static std::unique_ptr<Module> emit_function(
             }
             if (!found) {
                 if (VN) {
-                    Value *undef = VN->getType() == T_prjlvalue ?
-                        (llvm::Value*)ConstantPointerNull::get(cast<PointerType>(T_prjlvalue)) :
-                        (llvm::Value*)UndefValue::get(VN->getType());
-                    VN->addIncoming(undef, pred);
+                    if (!VNUndef) {
+                        // While we're guaranteed that this value will not be looked at dynamically,
+                        // we're not guaranteed that it won't be moved around. For things allocated on the
+                        // stack, that generally involves memcpying them from one stack slot to another,
+                        // so we need to make sure that such values are legal to be copied from.
+                        if (VN->getType() == T_prjlvalue) {
+                            VNUndef =  (llvm::Value*)ConstantPointerNull::get(cast<PointerType>(T_prjlvalue));
+                        } else if (VN->getType()->isPointerTy()) {
+                            PointerType *PTy = cast<PointerType>(VN->getType());
+                            VNUndef = new AllocaInst(PTy->getElementType(),
```

Comment:
I don't think we do / can correctly set the element type in many cases (especially with union-alloca)

Reply:
Ah, fair point, we can just get the size from the union type. Alternatively, we could just emit one alloca big enough for all phis at the top.